File: internal/util/redact_test.go
package util

import "testing"

func TestRedactAny(t *testing.T) {
	in := map[string]any{
		"apiKey": "abc",
		"nested": map[string]any{
			"token":     "t1",
			"safeField": "ok",
		},
		"list": []any{map[string]any{"password": "p"}},
	}
	out := RedactAny(in).(map[string]any)
	if out["apiKey"] != "***REDACTED***" {
		t.Fatalf("apiKey should be redacted")
	}
	nested := out["nested"].(map[string]any)
	if nested["token"] != "***REDACTED***" {
		t.Fatalf("nested token should be redacted")
	}
	if nested["safeField"] != "ok" {
		t.Fatalf("safe field should be unchanged")
	}
}

File: frontend/tests/unit/redact.test.ts
import { describe, expect, it } from 'vitest';
import { redactAny, shouldRedactKey } from '../../src/engine/util/redact';

describe('redact', () => {
  it('detects secret keys by token', () => {
    expect(shouldRedactKey('apiKey')).toBe(true);
    expect(shouldRedactKey('secret_access_key')).toBe(true);
    expect(shouldRedactKey('displayName')).toBe(false);
  });

  it('redacts nested sensitive keys', () => {
    const input = {
      apiKey: 'abc',
      nested: {
        password: 'pwd',
        keep: 'x',
      },
    };

    const output = redactAny(input) as Record<string, unknown>;
    expect(output.apiKey).toBe('***REDACTED***');
    expect((output.nested as Record<string, unknown>).password).toBe('***REDACTED***');
    expect((output.nested as Record<string, unknown>).keep).toBe('x');
  });
});
